### DISCUSSION TITLE
[Question][Hw6] Coverage.read doesn't close the input channel of coverage file?

### DISCUSSION THREAD

**Discussion by m-spitfire** [prosyslab-classroom/cs348-information-security#721] [Q&A]
Name: Murad Bashirov

Hello. How does autograder calculates coverage in json parser? I have num of unique crashes: 5, but covered lines: 0.

**Comment by m-spitfire** [comment]
I think I found the reason. When I do `timeout x` at some point `Coverage.read` gives EMFILE error, which in the end I end up with too many open files. I don't get the file descriptor from `Coverage.read` to close the file, how can I deal with this issue?

**Comment by m-spitfire** [comment]
This diff solves the issue:
```diff
diff --git a/src/coverage.ml b/src/coverage.ml
index 9d17c0e..673328a 100644
--- a/src/coverage.ml
+++ b/src/coverage.ml
@@ -22,6 +22,6 @@ let read coverage_file =
         line |> List.tl |> List.hd |> String.trim |> int_of_string
       in
       (if count > 0 then add line_number coverage else coverage) |> loop ic
-    with End_of_file -> coverage
+    with End_of_file -> close_in ic; coverage
   in
   loop ic empty
```
but as I can't modify the `src/coverage.ml` file, I'm not sure what to do.

**Comment by yeonhee-ryou** [comment]
We accumulates the coverage information in `*.gcov` file. Check if the files are generated well during the execution.

**Comment by m-spitfire** [comment]
@yeonhee-ryou my issue title is misleading now. I read that file every time I test it, but don't close it->it results in EMFILE error, which returns early with an error.

**Comment by yeonhee-ryou** [comment]
I'll check your issue and make a comment tomorrow

**Comment by m-spitfire** [comment]
solved it via implementing my own read coverage function. will keep open for the update.
Edit: update: #329
